OMCompiler/Compiler/SimCode/SerializeSparsityPattern.mo

```diff
@@ -0,0 +1,139 @@
+ * THIS OSMC PUBLIC LICENSE (OSMC-PL) VERSION 1.2.
+ * ANY USE, REPRODUCTION OR DISTRIBUTION OF THIS PROGRAM CONSTITUTES
+ * RECIPIENT'S ACCEPTANCE OF THE OSMC PUBLIC LICENSE OR THE GPL VERSION 3,
+ * ACCORDING TO RECIPIENTS CHOICE.
+ *
+ * The OpenModelica software and the Open Source Modelica
+ * Consortium (OSMC) Public License (OSMC-PL) are obtained
+ * from OSMC, either from the above address,
+ * from the URLs: http://www.ida.liu.se/projects/OpenModelica or
+ * http://www.openmodelica.org, and in the OpenModelica distribution.
+ * GNU version 3 is obtained from: http://www.gnu.org/copyleft/gpl.html.
+ *
+ * This program is distributed WITHOUT ANY WARRANTY; without
+ * even the implied warranty of  MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE, EXCEPT AS EXPRESSLY SET FORTH
+ * IN THE BY RECIPIENT SELECTED SUBSIDIARY LICENSE CONDITIONS OF OSMC-PL.
+ *
+ * See the full OSMC Public License conditions for more details.
+ *
+ */
+
+encapsulated package SerializeSparsityPattern
+
+import List;
+import SimCode;
+import Util;
+
+function serialize
+  input SimCode.SimCode code;
+  output String dummy = "";
+protected
+  array<Integer> columnPointers, rowIndices, columns;
+  String fname;
+algorithm
+  for jac in code.jacobianMatrices loop
+    if not listEmpty(jac.sparsity) then
+      fname := code.fileNamePrefix + "_Jac" + jac.matrixName + ".bin";
+      columnPointers := listArray(0 :: list(listLength(Util.tuple22(column)) for column in jac.sparsity));
+      rowIndices := listArray(List.flatten(list(Util.tuple22(column) for column in jac.sparsity)));
+      serializeJacobian(fname, arrayLength(columnPointers), arrayLength(rowIndices), columnPointers, rowIndices);
+      for color in jac.coloredCols loop
+        columns := listArray(color);
+        serializeColor(fname, arrayLength(columns), columns);
+      end for;
+    end if;
+  end for;
+end serialize;
+
+// *********************
+// write to binary stuff
+// *********************
+
+protected
+  function serializeJacobian
+    input String name;
+    input Integer numCols;
+    input Integer nnz;
+    input array<Integer> colPtrs;
+    input array<Integer> rowInds;
+  external "C" serializeJ(name, numCols, nnz, colPtrs, rowInds) annotation(Include="
+  extern FILE* omc_fopen(const char *filename, const char *mode);
+  extern size_t omc_fwrite(void *buffer, size_t size, size_t count, FILE *stream);
+
+  static void serializeJ(const char* name, int numCols, int nnz, modelica_metatype colPtrs, modelica_metatype rowInds)
+  {
+    unsigned int i, j;
+    size_t count;
+    FILE* pFile = omc_fopen(name, \"wb\");
+    if (pFile == NULL) {
+      throwStreamPrint(NULL, \"Could not open sparsity pattern file %s.\", name);
+    }
+
+    /* compute and write sparsePattern->leadindex */
+    j = 0;
+    for (i = 0; i < numCols; i++) {
+      j += (unsigned int) MMC_UNTAGFIXNUM(MMC_STRUCTDATA(colPtrs)[i]);
+      count = omc_fwrite(&j, sizeof(unsigned int), 1, pFile);
```

Comment:
Will this break cross-compiled FMUs? E.g. when compiling from a x86_64 Windows to a arm32 target?

Reply:
Good point. Can you check on Windows?